### PR TITLE
Simplification of SDK

### PR DESCRIPTION
**Pull request by edwarnicke**
This PR simplifies the SDK, by using the context.Context to pass
data between Endpoints in the chain.

By doing this, and a few other minor changes, the Endpoints used
in CompositeEndpoint just need to be normal implementers of

networkservice.NetworkServiceServer

Among other things, this means that CompositeEndpoints can be nested
within each other.

Note: Updates to sdk/README.md will be forthcoming.

Signed-off-by: Ed Warnicke <hagbard@gmail.com>

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: sdk/endpoint/client.go [hunk 1]
```diff
@@ -27,50 +27,47 @@ import (
...
+		if err != nil {
+			return nil, err
+		}
+	}
 
-	// TODO: check this. Hack??
-	outgoingConnection.GetMechanism().GetParameters()[connection.Workspace] = ""
-
-	cce.ioConnMap[incomingConnection.GetId()] = outgoingConnection
+	cce.ioConnMap[request.GetConnection().GetId()] = outgoingConnection
```
Comment (onobrod) [inline sdk/endpoint/client.go]: Should we save outgoingConnection into the map just after we created it to be able to close it even if next request will fail?

File: sdk/vppagent/commit.go [hunk 2]
```diff
@@ -23,99 +23,81 @@ const (
...
-		return nil, err
-	}
-
-	incomingConnection, err := f.GetNext().Request(ctx, request)
-	if err != nil {
-		return nil, err
+// Provides/Consumes from ctx context.Context:
+//     VppAgentConfig
+//	   Next
+func (f *Commit) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
```
Comment (onobrod) [inline sdk/vppagent/commit.go]: Please rename the receiver names to `c` also

File: sdk/vppagent/commit.go [hunk 3]
```diff
@@ -23,99 +23,81 @@ const (
...
+	if endpoint.Next(ctx) != nil {
+		return endpoint.Next(ctx).Close(ctx, connection)
 	}
 	return &empty.Empty{}, nil
 }
 
 // Name returns the composite name
-func (f *Flush) Name() string {
+func (f *Commit) Name() string {
 	return "flush"
```
Comment (onobrod) [inline sdk/vppagent/commit.go]: Please fix this name also

File: sdk/vppagent/commit.go [hunk 4]
```diff
@@ -23,99 +23,81 @@ const (
...
 // Close implements the close handler
-func (f *Flush) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
-	connectionData, err := getConnectionData(f.GetNext(), connection, false)
-	if err != nil {
-		return &empty.Empty{}, err
-	}
+// Provides/Consumes from ctx context.Context:
+//     VppAgentConfig
+//	   Next
+func (f *Commit) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
```
Comment (onobrod) [inline sdk/vppagent/commit.go]: Seems all /sdk/vppagent composites (except `Commit`) have almost identical Request() and Close() methods. Can we fix this somehow? Maybe can we store flushed data in a map inside `Commit` and just get it from it on Close()?

**Comment by onobrod** [comment]
I like that child endpoints of a composite endpoint are called in direct order now, it looks more clear and seems it is exactly what you asking about in #1282

File: sdk/vppagent/commit.go (hunk 2 above)
Comment (edwarnicke) [reply sdk/vppagent/commit.go]: Good catch :)  Will do :)

File: sdk/vppagent/commit.go (hunk 4 above)
Comment (edwarnicke) [reply sdk/vppagent/commit.go]: Interesting idea :)

**Comment by edwarnicke** [comment]
@onobrod Yes, but what you wrote was congruent with the SDK as it was at the time.  You did well.  It just brought a bunch of other things to my attention which I'm hacking at :)

File: sdk/endpoint/composite.go [hunk 5]
```diff
@@ -19,82 +19,71 @@ import (
...
 	GrpcServer *grpc.Server
 }
 
-// ChainedEndpoint is the basic endpoint composition interface
-type ChainedEndpoint interface {
-	networkservice.NetworkServiceServer
-	Name() string
-	Init(context *InitContext) error
-	GetNext() ChainedEndpoint
-	GetOpaque(interface{}) interface{}
```
Comment (onobrod) [inline sdk/endpoint/composite.go]: Need to update sdk documentation here: https://github.com/networkservicemesh/networkservicemesh/blob/master/sdk/README.md#writing-a-chainedendpoint

**Comment by haiodo** [comment]
Work is moved into #1524

**Pull request closed without merging**